File: elixir/lib/calculator.ex
defmodule Calculator do
  @moduledoc """
  Scientific Calculator in Elixir
  """

  @constants %{
    "pi" => :math.pi(),
    "e" => :math.exp(1),
    "c" => 299_792_458.0,
    "g" => 9.80665,
    "h" => 6.62607015e-34
  }

  defstruct memory: 0.0, history: [], radian_mode: false

  def new, do: %Calculator{}

  def to_radians(deg), do: deg * :math.pi() / 180.0
  def to_degrees(rad), do: rad * 180.0 / :math.pi()

  def get_angle(%Calculator{radian_mode: true}, x), do: x
  def get_angle(%Calculator{radian_mode: false}, x), do: to_radians(x)

  def from_angle(%Calculator{radian_mode: true}, r), do: r
  def from_angle(%Calculator{radian_mode: false}, r), do: to_degrees(r)

  def calculate(a, b, op) do
    result = case op do
      "+" -> {:ok, a + b}
      "-" -> {:ok, a - b}
      "*" -> {:ok, a * b}
      "/" when b == 0 -> {:error, "Division by zero"}
      "/" -> {:ok, a / b}
      "%" when b == 0 -> {:error, "Modulo by zero"}
      "%" -> {:ok, :math.fmod(a, b)}
      "^" ->
        try do
          {:ok, :math.pow(a, b)}
        rescue
          ArithmeticError -> {:error, "Invalid operation (result is undefined)"}
        end
      _ -> {:error, "Unknown operator: #{op}"}
    end
    # Check for NaN or Infinity
    check_result(result)
  end

  defp check_result({:error, _} = err), do: err
  defp check_result({:ok, r}) do
    cond do
      not is_number(r) or r != r -> {:error, "Invalid operation (result is undefined)"}
      is_float(r) and (r == :inf or r == :"-inf") -> {:error, "Overflow (result too large)"}
      is_float(r) and abs(r) > 1.0e308 -> {:error, "Overflow (result too large)"}
      true -> {:ok, r}
    end
  end

  def factorial(n) when n < 0 or n != trunc(n) do
    {:error, "Factorial requires non-negative integer"}
  end
  def factorial(n) when n > 170 do
    {:error, "Factorial overflow (n > 170)"}
  end
  def factorial(n) when n <= 1, do: {:ok, 1.0}
  def factorial(n) do
    result = Enum.reduce(2..trunc(n), 1.0, fn i, acc -> acc * i end)
    {:ok, result}
  end

  def unary_function(calc, func, x) do
    case String.downcase(func) do
      "sin" -> {:ok, :math.sin(get_angle(calc, x))}
      "cos" -> {:ok, :math.cos(get_angle(calc, x))}
      "tan" ->
        angle = get_angle(calc, x)
        if abs(:math.cos(angle)) < 1.0e-9 do
          {:error, "Undefined tan (angle near 90°)"}
        else
          {:ok, :math.tan(angle)}
        end
      "cot" ->
        angle = get_angle(calc, x)
        if abs(:math.sin(angle)) < 1.0e-9 do
          {:error, "Undefined cot (angle near 0°)"}
        else
          {:ok, 1.0 / :math.tan(angle)}
        end
      "sec" ->
        angle = get_angle(calc, x)
        if abs(:math.cos(angle)) < 1.0e-9 do
          {:error, "Undefined sec (angle near 90°)"}
        else
          {:ok, 1.0 / :math.cos(angle)}
        end
      "csc" ->
        angle = get_angle(calc, x)
        if abs(:math.sin(angle)) < 1.0e-9 do
          {:error, "Undefined csc (angle near 0°)"}
        else
          {:ok, 1.0 / :math.sin(angle)}
        end
      "asin" when x < -1 or x > 1 -> {:error, "asin domain is [-1, 1]"}
      "asin" -> {:ok, from_angle(calc, :math.asin(x))}
      "acos" when x < -1 or x > 1 -> {:error, "acos domain is [-1, 1]"}
      "acos" -> {:ok, from_angle(calc, :math.acos(x))}
      "atan" -> {:ok, from_angle(calc, :math.atan(x))}
      "sinh" -> {:ok, :math.sinh(x)}
      "cosh" -> {:ok, :math.cosh(x)}
      "tanh" -> {:ok, :math.tanh(x)}
      "asinh" -> {:ok, :math.asinh(x)}
      "acosh" when x < 1 -> {:error, "acosh domain is [1, ∞)"}
      "acosh" -> {:ok, :math.acosh(x)}
      "atanh" when x <= -1 or x >= 1 -> {:error, "atanh domain is (-1, 1)"}
      "atanh" -> {:ok, :math.atanh(x)}
      "sqrt" when x < 0 -> {:error, "Cannot take square root of negative number"}
      "sqrt" -> {:ok, :math.sqrt(x)}
      "cbrt" -> {:ok, sign(x) * :math.pow(abs(x), 1/3)}
      "ln" when x <= 0 -> {:error, "ln domain is (0, ∞)"}
      "ln" -> {:ok, :math.log(x)}
      "log" when x <= 0 -> {:error, "log domain is (0, ∞)"}
      "log" -> {:ok, :math.log10(x)}
      "exp" ->
        if x > 709 do
          {:error, "Overflow (exp result too large)"}
        else
          {:ok, :math.exp(x)}
        end
      "abs" -> {:ok, abs(x)}
      "factorial" -> factorial(x)
      _ -> {:error, "Unknown function: #{func}"}
    end
  end

  defp sign(x) when x < 0, do: -1
  defp sign(_), do: 1

  def convert_units(value, from, to) do
    case {String.downcase(from), String.downcase(to)} do
      {"m", "cm"} -> {:ok, value * 100}
      {"cm", "m"} -> {:ok, value / 100}
      {"km", "m"} -> {:ok, value * 1000}
      {"m", "km"} -> {:ok, value / 1000}
      {"deg", "rad"} -> {:ok, to_radians(value)}
      {"rad", "deg"} -> {:ok, to_degrees(value)}
      _ -> {:error, "Unsupported conversion"}
    end
  end

  def statistics([]), do: {0.0, 0.0}
  def statistics(data) do
    n = length(data)
    mean = Enum.sum(data) / n
    variance = Enum.sum(Enum.map(data, fn x -> :math.pow(x - mean, 2) end)) / n
    {mean, :math.sqrt(variance)}
  end

  def add_to_history(calc, entry) do
    history = [entry | calc.history] |> Enum.take(10)
    %{calc | history: history}
  end

  def toggle_mode(calc) do
    %{calc | radian_mode: not calc.radian_mode}
  end

  def reset(_calc), do: new()

  def get_constants, do: @constants
end

defmodule Calculator.CLI do
  def main(_args) do
    IO.puts("Starting Scientific Calculator (Elixir)...")
    loop(Calculator.new())
  end

  defp loop(calc) do
    show_menu()
    choice = prompt("Select option: ") |> String.trim()
    
    case choice do
      "1" -> handle_basic_op(calc) |> loop()
      "2" -> handle_unary(calc, "sqrt, cbrt, ln, log, exp, abs, factorial") |> loop()
      "3" -> handle_unary(calc, "sin, cos, tan") |> loop()
      "4" -> handle_unary(calc, "asin, acos, atan") |> loop()
      "5" -> handle_unary(calc, "sinh, cosh, tanh") |> loop()
      "6" -> handle_conversion(calc) |> loop()
      "7" -> handle_statistics(calc) |> loop()
      "8" -> handle_memory_add(calc) |> loop()
      "9" -> handle_memory_sub(calc) |> loop()
      "10" -> 
        IO.puts("Memory: #{calc.memory}")
        loop(calc)
      "11" ->
        IO.puts("Memory cleared.")
        loop(%{calc | memory: 0.0})
      "12" ->
        new_calc = Calculator.toggle_mode(calc)
        mode = if new_calc.radian_mode, do: "Radians", else: "Degrees"
        IO.puts("Mode: #{mode}")
        loop(new_calc)
      "13" ->
        show_history(calc)
        loop(calc)
      "14" ->
        IO.puts("Calculator reset.")
        loop(Calculator.new())
      "15" ->
        IO.puts("Goodbye!")
      _ ->
        IO.puts("Invalid option.")
        loop(calc)
    end
  end

  defp show_menu do
    IO.puts("\n=== Scientific Calculator (Elixir) ===")
    IO.puts("1.  Basic Operation (+, -, *, /, %, ^)")
    IO.puts("2.  Unary Function (sqrt, cbrt, ln, log, exp, abs, factorial)")
    IO.puts("3.  Trigonometry (sin, cos, tan)")
    IO.puts("4.  Inverse Trig (asin, acos, atan)")
    IO.puts("5.  Hyperbolic (sinh, cosh, tanh)")
    IO.puts("6.  Unit Conversion")
    IO.puts("7.  Statistics (mean, std dev)")
    IO.puts("8.  Memory Add (M+)")
    IO.puts("9.  Memory Subtract (M-)")
    IO.puts("10. Memory Recall (MR)")
    IO.puts("11. Memory Clear (MC)")
    IO.puts("12. Toggle Mode (Deg/Rad)")
    IO.puts("13. Show History")
    IO.puts("14. Reset")
    IO.puts("15. Exit")
  end

  defp prompt(msg) do
    IO.write(msg)
    IO.read(:stdio, :line) |> String.trim()
  end

  defp prompt_float(msg) do
    case Float.parse(prompt(msg)) do
      {num, _} -> num
      :error -> 0.0
    end
  end

  defp handle_basic_op(calc) do
    a = prompt_float("Enter first number: ")
    op = prompt("Enter operator (+, -, *, /, %, ^): ")
    b = prompt_float("Enter second number: ")
    
    case Calculator.calculate(a, b, op) do
      {:ok, result} ->
        IO.puts("Result: #{result}")
        entry = "#{a} #{op} #{b} = #{result}"
        Calculator.add_to_history(calc, entry)
      {:error, err} ->
        IO.puts("Error: #{err}")
        calc
    end
  end

  defp handle_unary(calc, funcs) do
    func = prompt("Enter function (#{funcs}): ")
    x = prompt_float("Enter value: ")
    
    case Calculator.unary_function(calc, func, x) do
      {:ok, result} ->
        unit = if String.starts_with?(func, "a") and func != "abs", 
               do: if(calc.radian_mode, do: " rad", else: "°"), 
               else: ""
        IO.puts("Result: #{result}#{unit}")
        entry = "#{func}(#{x}) = #{result}"
        Calculator.add_to_history(calc, entry)
      {:error, err} ->
        IO.puts("Error: #{err}")
        calc
    end
  end

  defp handle_conversion(calc) do
    value = prompt_float("Enter value: ")
    from = prompt("From unit (m, cm, km, deg, rad): ")
    to = prompt("To unit: ")
    
    case Calculator.convert_units(value, from, to) do
      {:ok, result} ->
        IO.puts("Result: #{result} #{to}")
        entry = "#{value} #{from} = #{result} #{to}"
        Calculator.add_to_history(calc, entry)
      {:error, err} ->
        IO.puts("Error: #{err}")
        calc
    end
  end

  defp handle_statistics(calc) do
    input = prompt("Enter numbers (space-separated): ")
    data = input
           |> String.split()
           |> Enum.map(&Float.parse/1)
           |> Enum.filter(fn x -> x != :error end)
           |> Enum.map(fn {n, _} -> n end)
    
    {mean, std_dev} = Calculator.statistics(data)
    IO.puts("Mean: #{mean}, Std Dev: #{std_dev}")
    entry = "Stats: mean=#{mean}, std=#{std_dev}"
    Calculator.add_to_history(calc, entry)
  end

  defp handle_memory_add(calc) do
    value = prompt_float("Value to add: ")
    new_calc = %{calc | memory: calc.memory + value}
    IO.puts("Memory: #{new_calc.memory}")
    new_calc
  end

  defp handle_memory_sub(calc) do
    value = prompt_float("Value to subtract: ")
    new_calc = %{calc | memory: calc.memory - value}
    IO.puts("Memory: #{new_calc.memory}")
    new_calc
  end

  defp show_history(calc) do
    if calc.history == [] do
      IO.puts("History is empty.")
    else
      IO.puts("\n=== History ===")
      calc.history |> Enum.reverse() |> Enum.each(&IO.puts/1)
    end
  end
end

File: haskell/Calculator.hs
module Main where

import Text.Read (readMaybe)
import Data.Char (toLower)
import Control.Monad (when)
import Data.IORef
import System.IO (hFlush, stdout)

data CalcState = CalcState
    { memory     :: Double
    , history    :: [String]
    , radianMode :: Bool
    } deriving (Show)

initialState :: CalcState
initialState = CalcState 0.0 [] False

constants :: [(String, Double)]
constants = 
    [ ("pi", pi)
    , ("e", exp 1)
    , ("c", 299792458.0)
    , ("g", 9.80665)
    , ("h", 6.62607015e-34)
    ]

toRadians :: Double -> Double
toRadians deg = deg * pi / 180.0

toDegrees :: Double -> Double
toDegrees rad = rad * 180.0 / pi

getAngle :: Bool -> Double -> Double
getAngle True x  = x
getAngle False x = toRadians x

fromAngle :: Bool -> Double -> Double
fromAngle True r  = r
fromAngle False r = toDegrees r

calculate :: Double -> Double -> Char -> Either String Double
calculate a b op = checkResult $ case op of
    '+' -> Right (a + b)
    '-' -> Right (a - b)
    '*' -> Right (a * b)
    '/' -> if b == 0 then Left "Division by zero" else Right (a / b)
    '%' -> if b == 0 then Left "Modulo by zero" else Right (a `mod'` b)
    '^' -> Right (a ** b)
    _   -> Left $ "Unknown operator: " ++ [op]
  where
    mod' x y = x - y * fromIntegral (floor (x / y))
    checkResult (Left err) = Left err
    checkResult (Right r)
        | isNaN r = Left "Invalid operation (result is undefined)"
        | isInfinite r = Left "Overflow (result too large)"
        | otherwise = Right r

factorial :: Double -> Either String Double
factorial n
    | n < 0 || n /= fromIntegral (floor n) = Left "Factorial requires non-negative integer"
    | n > 170 = Left "Factorial overflow (n > 170)"
    | n <= 1 = Right 1.0
    | otherwise = Right $ product [1.0 .. n]

unaryFunction :: Bool -> String -> Double -> Either String Double
unaryFunction radMode func x = case map toLower func of
    "sin"       -> Right $ sin (getAngle radMode x)
    "cos"       -> Right $ cos (getAngle radMode x)
    "tan"       -> let angle = getAngle radMode x
                   in if abs (cos angle) < 1e-9
                      then Left "Undefined tan (angle near 90°)"
                      else Right $ tan angle
    "cot"       -> let angle = getAngle radMode x
                   in if abs (sin angle) < 1e-9
                      then Left "Undefined cot (angle near 0°)"
                      else Right $ 1 / tan angle
    "sec"       -> let angle = getAngle radMode x
                   in if abs (cos angle) < 1e-9
                      then Left "Undefined sec (angle near 90°)"
                      else Right $ 1 / cos angle
    "csc"       -> let angle = getAngle radMode x
                   in if abs (sin angle) < 1e-9
                      then Left "Undefined csc (angle near 0°)"
                      else Right $ 1 / sin angle
    "asin"      -> if x < -1 || x > 1 
                   then Left "asin domain is [-1, 1]"
                   else Right $ fromAngle radMode (asin x)
    "acos"      -> if x < -1 || x > 1
                   then Left "acos domain is [-1, 1]"
                   else Right $ fromAngle radMode (acos x)
    "atan"      -> Right $ fromAngle radMode (atan x)
    "sinh"      -> Right $ sinh x
    "cosh"      -> Right $ cosh x
    "tanh"      -> Right $ tanh x
    "asinh"     -> Right $ asinh x
    "acosh"     -> if x < 1
                   then Left "acosh domain is [1, ∞)"
                   else Right $ acosh x
    "atanh"     -> if x <= -1 || x >= 1
                   then Left "atanh domain is (-1, 1)"
                   else Right $ atanh x
    "sqrt"      -> if x < 0
                   then Left "Cannot take square root of negative number"
                   else Right $ sqrt x
    "cbrt"      -> Right $ signum x * (abs x ** (1/3))
    "ln"        -> if x <= 0
                   then Left "ln domain is (0, ∞)"
                   else Right $ log x
    "log"       -> if x <= 0
                   then Left "log domain is (0, ∞)"
                   else Right $ logBase 10 x
    "exp"       -> let result = exp x
                   in if isInfinite result
                      then Left "Overflow (exp result too large)"
                      else Right result
    "abs"       -> Right $ abs x
    "factorial" -> factorial x
    _           -> Left $ "Unknown function: " ++ func

convertUnits :: Double -> String -> String -> Either String Double
convertUnits value from to = case (map toLower from, map toLower to) of
    ("m", "cm")   -> Right $ value * 100
    ("cm", "m")   -> Right $ value / 100
    ("km", "m")   -> Right $ value * 1000
    ("m", "km")   -> Right $ value / 1000
    ("deg", "rad") -> Right $ toRadians value
    ("rad", "deg") -> Right $ toDegrees value
    _             -> Left "Unsupported conversion"

statistics :: [Double] -> (Double, Double)
statistics [] = (0.0, 0.0)
statistics xs = (mean, stdDev)
  where
    n = fromIntegral $ length xs
    mean = sum xs / n
    variance = sum (map (\x -> (x - mean) ** 2) xs) / n
    stdDev = sqrt variance

addToHistory :: String -> CalcState -> CalcState
addToHistory entry state = 
    state { history = take 10 (entry : history state) }

showHistory :: CalcState -> IO ()
showHistory state = 
    if null (history state)
    then putStrLn "History is empty."
    else do
        putStrLn "\n=== History ==="
        mapM_ putStrLn (reverse $ history state)

toggleMode :: CalcState -> CalcState
toggleMode state = state { radianMode = not (radianMode state) }

resetCalc :: CalcState -> CalcState
resetCalc _ = initialState

showMenu :: IO ()
showMenu = do
    putStrLn "\n=== Scientific Calculator (Haskell) ==="
    putStrLn "1.  Basic Operation (+, -, *, /, %, ^)"
    putStrLn "2.  Unary Function (sqrt, cbrt, ln, log, exp, abs, factorial)"
    putStrLn "3.  Trigonometry (sin, cos, tan)"
    putStrLn "4.  Inverse Trig (asin, acos, atan)"
    putStrLn "5.  Hyperbolic (sinh, cosh, tanh)"
    putStrLn "6.  Unit Conversion"
    putStrLn "7.  Statistics (mean, std dev)"
    putStrLn "8.  Memory Add (M+)"
    putStrLn "9.  Memory Subtract (M-)"
    putStrLn "10. Memory Recall (MR)"
    putStrLn "11. Memory Clear (MC)"
    putStrLn "12. Toggle Mode (Deg/Rad)"
    putStrLn "13. Show History"
    putStrLn "14. Reset"
    putStrLn "15. Exit"

prompt :: String -> IO String
prompt msg = do
    putStr msg
    hFlush stdout
    getLine

promptDouble :: String -> IO Double
promptDouble msg = do
    input <- prompt msg
    return $ maybe 0.0 id (readMaybe input)

main :: IO ()
main = do
    stateRef <- newIORef initialState
    loop stateRef
  where
    loop stateRef = do
        showMenu
        choice <- prompt "\nSelect option: "
        state <- readIORef stateRef
        
        case choice of
            "1" -> do
                a <- promptDouble "Enter first number: "
                opStr <- prompt "Enter operator (+, -, *, /, %, ^): "
                let op = if null opStr then '+' else head opStr
                b <- promptDouble "Enter second number: "
                case calculate a b op of
                    Right result -> do
                        putStrLn $ "Result: " ++ show result
                        let entry = show a ++ " " ++ [op] ++ " " ++ show b ++ " = " ++ show result
                        modifyIORef stateRef (addToHistory entry)
                    Left err -> putStrLn $ "Error: " ++ err
                loop stateRef
                
            "2" -> do
                func <- prompt "Enter function (sqrt, cbrt, ln, log, exp, abs, factorial): "
                x <- promptDouble "Enter value: "
                case unaryFunction (radianMode state) func x of
                    Right result -> do
                        putStrLn $ "Result: " ++ show result
                        let entry = func ++ "(" ++ show x ++ ") = " ++ show result
                        modifyIORef stateRef (addToHistory entry)
                    Left err -> putStrLn $ "Error: " ++ err
                loop stateRef
                
            "3" -> do
                func <- prompt "Enter function (sin, cos, tan): "
                x <- promptDouble "Enter angle: "
                case unaryFunction (radianMode state) func x of
                    Right result -> do
                        putStrLn $ "Result: " ++ show result
                        let entry = func ++ "(" ++ show x ++ ") = " ++ show result
                        modifyIORef stateRef (addToHistory entry)
                    Left err -> putStrLn $ "Error: " ++ err
                loop stateRef
                
            "4" -> do
                func <- prompt "Enter function (asin, acos, atan): "
                x <- promptDouble "Enter value: "
                case unaryFunction (radianMode state) func x of
                    Right result -> do
                        let unit = if radianMode state then " rad" else "°"
                        putStrLn $ "Result: " ++ show result ++ unit
                        let entry = func ++ "(" ++ show x ++ ") = " ++ show result
                        modifyIORef stateRef (addToHistory entry)
                    Left err -> putStrLn $ "Error: " ++ err
                loop stateRef
                
            "5" -> do
                func <- prompt "Enter function (sinh, cosh, tanh): "
                x <- promptDouble "Enter value: "
                case unaryFunction (radianMode state) func x of
                    Right result -> do
                        putStrLn $ "Result: " ++ show result
                        let entry = func ++ "(" ++ show x ++ ") = " ++ show result
                        modifyIORef stateRef (addToHistory entry)
                    Left err -> putStrLn $ "Error: " ++ err
                loop stateRef
                
            "6" -> do
                value <- promptDouble "Enter value: "
                from <- prompt "From unit (m, cm, km, deg, rad): "
                to <- prompt "To unit: "
                case convertUnits value from to of
                    Right result -> do
                        putStrLn $ "Result: " ++ show result ++ " " ++ to
                        let entry = show value ++ " " ++ from ++ " = " ++ show result ++ " " ++ to
                        modifyIORef stateRef (addToHistory entry)
                    Left err -> putStrLn $ "Error: " ++ err
                loop stateRef
                
            "7" -> do
                input <- prompt "Enter numbers (space-separated): "
                let nums = mapMaybe readMaybe (words input) :: [Double]
                let (mean, stdDev) = statistics nums
                putStrLn $ "Mean: " ++ show mean ++ ", Std Dev: " ++ show stdDev
                let entry = "Stats: mean=" ++ show mean ++ ", std=" ++ show stdDev
                modifyIORef stateRef (addToHistory entry)
                loop stateRef
                
            "8" -> do
                value <- promptDouble "Value to add: "
                modifyIORef stateRef (\s -> s { memory = memory s + value })
                newState <- readIORef stateRef
                putStrLn $ "Memory: " ++ show (memory newState)
                loop stateRef
                
            "9" -> do
                value <- promptDouble "Value to subtract: "
                modifyIORef stateRef (\s -> s { memory = memory s - value })
                newState <- readIORef stateRef
                putStrLn $ "Memory: " ++ show (memory newState)
                loop stateRef
                
            "10" -> do
                putStrLn $ "Memory: " ++ show (memory state)
                loop stateRef
                
            "11" -> do
                modifyIORef stateRef (\s -> s { memory = 0.0 })
                putStrLn "Memory cleared."
                loop stateRef
                
            "12" -> do
                modifyIORef stateRef toggleMode
                newState <- readIORef stateRef
                putStrLn $ "Mode: " ++ if radianMode newState then "Radians" else "Degrees"
                loop stateRef
                
            "13" -> do
                showHistory state
                loop stateRef
                
            "14" -> do
                writeIORef stateRef initialState
                putStrLn "Calculator reset."
                loop stateRef
                
            "15" -> putStrLn "Goodbye!"
            
            _ -> do
                putStrLn "Invalid option."
                loop stateRef

mapMaybe :: (a -> Maybe b) -> [a] -> [b]
mapMaybe _ [] = []
mapMaybe f (x:xs) = case f x of
    Just y  -> y : mapMaybe f xs
    Nothing -> mapMaybe f xs
